### PR TITLE
trilinos: extend TPetra explicit instantiations and fix missing Teuch…

### PR DESCRIPTION
**Pull request by davydden**
…os counterparts

@jrood-nrel since you added `instantiate` variants in https://github.com/spack/spack/pull/4428 , please have a look at this PR.

### CODE REVIEW

File: var/spack/repos/builtin/packages/trilinos/package.py [hunk 1]
```diff
@@ -79,10 +79,12 @@ class Trilinos(CMakePackage):
     # Build options
     variant('fortran',      default=True,
             description='Compile with Fortran support')
+    variant('explicit_instantiation',  default=False,
+            description='Enable explicit instantiation')
     variant('instantiate',  default=True,
-            description='Compile with explicit instantiation')
+            description='Additional TPetra instantiations')
```
Comment (aprokop) [inline var/spack/repos/builtin/packages/trilinos/package.py]: The proper spelling is `Tpetra`, not `TPetra`.

File: var/spack/repos/builtin/packages/trilinos/package.py [hunk 2]
```diff
@@ -79,10 +79,12 @@ class Trilinos(CMakePackage):
     # Build options
     variant('fortran',      default=True,
             description='Compile with Fortran support')
+    variant('explicit_instantiation',  default=False,
+            description='Enable explicit instantiation')
     variant('instantiate',  default=True,
```
Comment (aprokop) [inline var/spack/repos/builtin/packages/trilinos/package.py]: I would rename this to be clear what it does. `extra_tpetra_instantiations`?

File: var/spack/repos/builtin/packages/trilinos/package.py [hunk 3]
```diff
@@ -678,20 +680,41 @@ def cmake_args(self):
...
+                    ),
+                    '-DTpetra_INST_COMPLEX_FLOAT=%s' % (
+                        'ON' if '+instantiate_cmplx' in spec else 'OFF'
+                    ),
+                    '-DTpetra_INST_FLOAT=ON'
                 ])
 
+                if '+teuchos' in spec:
+                    options.extend([
+                        '-DTeuchos_ENABLE_COMPLEX=%s' % (
```
Comment (aprokop) [inline var/spack/repos/builtin/packages/trilinos/package.py]: If I understand it right, `Teuchos_ENABLE_COMPLEX` has nothing to do with instantiations but rather enabling complex numbers. You can still use complex even without explicit instantiations.

File: var/spack/repos/builtin/packages/trilinos/package.py [hunk 4]
```diff
@@ -678,20 +680,41 @@ def cmake_args(self):
                     '-DTrilinos_ENABLE_Fortran=ON'
                 ])
 
-        # Explicit instantiation
-        if '+instantiate' in spec:
+        # Instantiation in Tpetra and Teuchos
+        # As of 12.12.1, DTrilinos_ENABLE_EXPLICIT_INSTANTIATION does not really build with
+        # extra instantiations of Tpetra, see
+        # https://github.com/trilinos/Trilinos/issues/3566 and
+        # https://github.com/trilinos/Trilinos/issues/3566
```
Comment (aprokop) [inline var/spack/repos/builtin/packages/trilinos/package.py]: The link to Trilinos issue is repeated. Did you mean a different one?

**Comment by davydden** [comment]
@aprokop please give it another look.

**Comment by jrood-nrel** [comment]
Thanks for making me aware of this. This looks good to me. I'm not tied to anything I did previously and this looks cleaner. It looks like I did this for Nalu, and I haven't looked at how it will affect it yet, but I will just update it accordingly once this is in.

**Comment by balay** [comment]
@davydden, any pertcular reason why superlu_dist is disabled here?

**Comment by davydden** [comment]
> @davydden, any pertcular reason why superlu_dist is disabled here?

I did not disable it anyhow, i just added a few conflicts that reflect the current state of affairs with float/complex support.

**Comment by davydden** [comment]
p.s. the default variants will still build fine with `superlu-dist`

**Comment by balay** [comment]
Ah Thanks! I was't looking at the 'float/complex' bits closely. All good for now.

**Pull request merged**